File: Program_Files/LinkedListRewrite.java
public class SingleLinkedList {

    private class Node {
        public Node link;
        public int contents;
    }

    private Node head;

    public SingleLinkedList() {
        head = null;
    }

    public void insertAtHead(int value);

    public void insertAtTail(int value);

    public int removeAtHead();

    public boolean isEmpty();

}

public class Stack2 extends SingleLinkedList {

    public Stack2() {
        super();
    }

    public void push(int value) {
        insertAtHead(value);
    }

    public int pop() {
        return removeAtHead();
    }

    public boolean isEmpty() {
        return super.isEmpty();
    }

}

public class Queue2 extends SingleLinkedList {

    public Queue2() {
        super();
    }

    public void enqueue(int value) {
        insertAtTail(value);
    }

    public int dequeue() {
        return removeAtHead();
    }

    public boolean isEmpty() {
        return super.isEmpty();
    }

}

File: Program_Files/LinkedListRewrite.rb
# Online Ruby Compiler (Interpreter)

# Base class SingleLinkedList
class SingleLinkedList
    class Node
      attr_accessor :link, :contents
    end
  
    def initialize
      @head = nil
    end
  
    def insert_at_head(value)
      node = Node.new
      node.contents = value
      node.link = @head
      @head = node
    end
  
    def insert_at_tail(value)
      node = Node.new
      node.contents = value
      if @head.nil?
        @head = node
      else
        tail_node = @head
        tail_node = tail_node.link until tail_node.link.nil?
        tail_node.link = node
      end
    end
  
    def remove_at_head
      raise 'List is empty' if @head.nil?
      node = @head
      @head = @head.link
      node.contents
    end
  
    def is_empty
      @head.nil?
    end
  end
  
  # Stack2 which extends SingleLinkedList
  class Stack2 < SingleLinkedList
    def push(value)
      insert_at_head(value)
    end
  
    def pop
      remove_at_head
    end
  
    def is_empty
      super
    end
  end
  
  # Queue2 which extends SingleLinkedList
  class Queue2 < SingleLinkedList
    def enqueue(value)
      insert_at_tail(value)
    end
  
    def dequeue
      remove_at_head
    end
  
    def is_empty
      super
    end
  end